File: parser/src/main/java/io/ticofab/androidgpxparser/parser/domain/Metadata.java
package io.ticofab.androidgpxparser.parser.domain;

import org.joda.time.DateTime;

public class Metadata {

    private final String mName;
    private final String mDesc;
    private final Author mAuthor;
    private final Copyright mCopyright;
    private final Link mLink;
    private final DateTime mTime;
    private final String mKeywords;
    private final Bounds mBounds;
    private final String mExtensions;

    private Metadata(Metadata.Builder builder) {
        mName = builder.mName;
        mDesc = builder.mDesc;
        mAuthor = builder.mAuthor;
        mCopyright = builder.mCopyright;
        mLink = builder.mLink;
        mTime = builder.mTime;
        mKeywords = builder.mKeywords;
        mBounds = builder.mBounds;
        mExtensions = builder.mExtensions;
    }

    public String getName() {
        return mName;
    }

    public String getDesc() {
        return mDesc;
    }

    public Author getAuthor() {
        return mAuthor;
    }

    public Copyright getCopyright() {
        return mCopyright;
    }

    public Link getLink() {
        return mLink;
    }

    public DateTime getTime() {
        return mTime;
    }

    public String getKeywords() {
        return mKeywords;
    }

    public Bounds getBounds() {
        return mBounds;
    }

    public String getExtensions() {
        return mExtensions;
    }

    public static class Builder {
        private String mName;
        private String mDesc;
        private Author mAuthor;
        private Copyright mCopyright;
        private Link mLink;
        private DateTime mTime;
        private String mKeywords;
        private Bounds mBounds;
        private String mExtensions;

        public Builder setName(String name) {
            mName = name;
            return this;
        }

        public Builder setDesc(String desc) {
            mDesc = desc;
            return this;
        }

        public Builder setAuthor(Author author) {
            mAuthor = author;
            return this;
        }

        public Builder setCopyright(Copyright copyright) {
            mCopyright = copyright;
            return this;
        }

        public Builder setLink(Link link) {
            mLink = link;
            return this;
        }

        public Builder setTime(DateTime time) {
            mTime = time;
            return this;
        }

        public Builder setKeywords(String keywords) {
            mKeywords = keywords;
            return this;
        }

        public Builder setBounds(Bounds bounds) {
            mBounds = bounds;
            return this;
        }

        public Metadata build() {
            return new Metadata(this);
        }


    }
}

File: parser/src/main/kotlin/io/ticofab/androidgpxparser/parser/domain/Metadata.kt
package io.ticofab.androidgpxparser.parser.domain

import org.joda.time.DateTime

class Metadata private constructor(builder: Builder) {
    val name = builder.name
    val desc = builder.desc
    val author = builder.author
    val copyright = builder.copyright
    val link = builder.link
    val time = builder.time
    val keywords = builder.keywords
    val bounds = builder.bounds
    val extensions = builder.extensions

    class Builder {
        var name: String? = null
            private set
        var desc: String? = null
            private set
        var author: Author? = null
            private set
        var copyright: Copyright? = null
            private set
        var link: Link? = null
            private set
        var time: DateTime? = null
            private set
        var keywords: String? = null
            private set
        var bounds: Bounds? = null
            private set
        val extensions: String? = null

        fun setName(name: String?) = apply { this.name = name }

        fun setDesc(desc: String?) = apply { this.desc = desc }

        fun setAuthor(author: Author?) = apply { this.author = author }

        fun setCopyright(copyright: Copyright?) = apply { this.copyright = copyright }

        fun setLink(link: Link?) = apply { this.link = link }

        fun setTime(time: DateTime?) = apply { this.time = time }

        fun setKeywords(keywords: String?) = apply { this.keywords = keywords }

        fun setBounds(bounds: Bounds?) = apply { this.bounds = bounds }

        fun build() = Metadata(this)
    }
}